pkg/volume/flocker/client.go

```diff
@@ -0,0 +1,391 @@
+	var configurations []configurationPayload
+	if err = json.NewDecoder(body).Decode(&configurations); err == nil {
+		for _, r := range configurations {
+			if r.Metadata.Name == name {
+				return r.DatasetID, nil
+			}
+		}
+		return "", errConfigurationNotFound
+	}
+	return "", err
+}
+
+// findPathInDatasetStatePayload returns the path of the given datasetID if it
+// was found in the state payload. In case the path is not found it returns an
+// error.
+func (c flockerClient) findPathInDatasetStatePayload(body io.ReadCloser, datasetID string) (path string, err error) {
+	var states []datasetStatePayload
+	if err = json.NewDecoder(body).Decode(&states); err == nil {
+		for _, s := range states {
+			if s.DatasetID == datasetID {
+				return s.Path, nil
+			}
+		}
+		return "", errStateNotFound
+	}
+	return "", err
+}
+
+// findUUIDInNodeStatePayload returns the UUID for the given host.
+func (c flockerClient) findUUIDInNodeStatePayload(body io.ReadCloser) (uuid string, err error) {
+	var states []nodeStatePayload
+	if err = json.NewDecoder(body).Decode(&states); err == nil {
+		for _, s := range states {
+			if s.Host == c.host {
+				return s.UUID, nil
+			}
+		}
+		return "", errStateNotFound
+	}
+	return "", err
+}
+
+// getDatasetState performs a get request to get the state of the given datasetID, if
+// something goes wrong or the datasetID was not found it returns an error.
+func (c flockerClient) getDatasetState(datasetID string) (*datasetState, error) {
+	resp, err := c.get(c.getURL("state/datasets"))
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	path, err := c.findPathInDatasetStatePayload(resp.Body, datasetID)
+	if err != nil {
+		return nil, err
+	}
+
+	return &datasetState{
+		DatasetID: datasetID,
+		Path:      path,
+	}, nil
+}
+
+/*
+ * CreateVolume creates a volume in Flocker, waits for it to be ready and
+ * returns the dataset id.
+ *
+ * This process is a little bit complex but follows this flow:
+ *
+ * 1) Get all the datasets
+ * 2) If a dataset with that name/dir is found, return its path
+ * 3) If not, create a new one (we will need the Primary UUID)
+ * 4) Wait until the dataset is ready or the timeout was reached
+ */
+func (c flockerClient) CreateVolume(dir string) (datasetID string, err error) {
+	// 1 & 2) Try to find the dataset if it was previously created
+	resp, err := c.get(c.getURL("configuration/datasets"))
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	if datasetID, err := c.findIDInConfigurationsPayload(resp.Body, dir); err == nil {
+		state, _ := c.getDatasetState(datasetID)
+		// If it fails here it can be recovered, this is why the error is ignored
+		if state.Path == dir {
+			return datasetID, nil
+		}
+	}
+
+	// 3) Create a new one if we get here
+	resp, err = c.get(c.getURL("state/nodes"))
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
```

Comment:
Why do you think so? `.get` calls `.request` and no one of those closes the body.